### PR TITLE
Commands suggestions

### LINKED ISSUE #163: Add commands suggestions to the CLI
I think it would be useful to have a suggestion commands when we enter `.` in the shell.
```
genji>. <suggestion of all commands>
```

### PR DESCRIPTION
**Pull request by tzzed**
This PR fixes #163. It suggests command when we enter `.` and suggests a command if the user make an input error.
```
genji>.hel
.hel is not a command. You meant: ".help"
genji>.foo 
unknown command ".foo". Enter ".help" for help.
```

### CODE REVIEW

File: cmd/genji/shell/command.go [hunk 1]
```diff
@@ -2,20 +2,94 @@ package shell
...
+
+// OrderedBy returns a Sorter that sorts using the less functions, in order.
+// Call its Sort method to sort the data.
+func OrderedBy(less ...lessFunc) *Commands {
+	return &Commands{
+		less: less,
+	}
+}
+
+// Len is part of sort.Interface.
```
Comment (asdine) [inline cmd/genji/shell/command.go:66]:
```suggestion
// Len implements sort.Interface.
```

File: cmd/genji/shell/command.go [hunk 2]
```diff
@@ -2,20 +2,94 @@ package shell
...
+func (c *Commands) Len() int {
+	return len(c.cmds)
+}
+
+// Swap is part of sort.Interface.
```
Comment (asdine) [inline cmd/genji/shell/command.go:71]:
```suggestion
// Swap implements sort.Interface.
```

File: cmd/genji/shell/command.go [hunk 3]
```diff
@@ -2,20 +2,94 @@ package shell
...
-	".help":                 "\t\t\tList all commands.",
+type Command struct {
+	Name    string
+	Options string
+	Help    string
+	Action  func(db *genji.DB, cmd []string) error
 }
 
-func runTablesCmd(db *genji.DB, cmd []string) error {
+type lessFunc func(a, b *Command) int
```
Comment (asdine) [inline cmd/genji/shell/command.go:21]: I don't think we need to go that far for dealing with sorting. There are barely 4 commands we can definitely chose the order manually. Also, I don't think alphabetical order is the best way to sort these commands, I think it would be better to group them by categories in the future.

File: cmd/genji/shell/command.go [hunk 4]
```diff
@@ -92,16 +165,100 @@ func runIndexesCmd(db *genji.DB, in []string) error {
...
 
-// runHelpCmd display all available dot commands.
-func runHelpCmd() error {
-	var keys []string
-	for k := range commands {
-		keys = append(keys, k)
+// displayHelp show all available commands.
+func displayHelp(db *genji.DB, cmd []string) error {
+	for _, c := range commands.cmds {
+		indent := 25 - len(c.getOptions()) - len(c.getName())
```
Comment (asdine) [inline cmd/genji/shell/command.go:171]: What does 25 mean?

File: cmd/genji/shell/command.go [hunk 5]
```diff
@@ -92,16 +165,100 @@ func runIndexesCmd(db *genji.DB, in []string) error {
...
 
-	sort.Strings(keys)
-	for _, k := range keys {
-		fmt.Printf("%s %s\n", k, commands[k])
+	return nil
+}
+
+// exitCommand exits this program.
+func exitCommand(db *genji.DB, cmd []string) error {
+	os.Exit(0)
```
Comment (asdine) [inline cmd/genji/shell/command.go:181]: Let's make sure the db is closed before exiting

File: cmd/genji/shell/command.go [hunk 6]
```diff
@@ -92,16 +165,100 @@ func runIndexesCmd(db *genji.DB, in []string) error {
...
+
+// getCommandSuggestion returns the suggested command.
+func getCommandSuggestion(in string) Command {
+	suggest := Command{
+		Action: func(db *genji.DB, cmd []string) error {
+			var list []string
+			for _, c := range commands.cmds {
+				name := c.getName()
+				d := levenshtein.ComputeDistance(name, in)
+				if d < 3 {
```
[the remark this answers is not in the document]
Comment (tzzed) [reply cmd/genji/shell/command.go:222]: You are right what do you thing about a distance smaller than the half of the command name. 
`if d < (len(c.Name) / 2)`

File: cmd/genji/shell/command.go (hunk 1 above)
Comment (tzzed) [reply cmd/genji/shell/command.go:66]: If we don't keep the sort I think is better to delete the `Len()` implementation.

File: cmd/genji/shell/command.go (hunk 2 above)
Comment (tzzed) [reply cmd/genji/shell/command.go:71]: same as above

File: cmd/genji/shell/command.go (hunk 3 above)
Comment (tzzed) [reply cmd/genji/shell/command.go:21]: I am agree and I like the "groupby" categories.

File: cmd/genji/shell/command.go (hunk 4 above)
Comment (tzzed) [reply cmd/genji/shell/command.go:171]: it's the initial number of spaces to keep all descriptions aligned when they are displayed.

File: cmd/genji/shell/shell.go [hunk 7]
```diff
@@ -370,6 +351,21 @@ func (sh *Shell) getAllTables() ([]string, error) {
 }
 
 func (sh *Shell) completer(in prompt.Document) []prompt.Suggest {
+	if strings.HasPrefix(in.Text, ".") {
+		expected := getAllCommands()
+		suggestions := make([]prompt.Suggest, len(expected))
+		for i, e := range expected {
+			suggestions[i].Text = e
+		}
```
Comment (asdine) [inline cmd/genji/shell/shell.go:359]: I think these 5 lines are the same regardless of the query, they should be done once and used here directly instead of being recreated every time a user runs a query.

File: cmd/genji/shell/shell.go [hunk 8]
```diff
@@ -217,32 +218,12 @@ func (sh *Shell) executeInput(in string) error {
 
 func (sh *Shell) runCommand(in string) error {
 	cmd := strings.Fields(in)
-	switch cmd[0] {
```
Comment (asdine) [inline cmd/genji/shell/shell.go:220]: I'm not sure why we don't keep this simple switch instead of doing globals and command types. Do you have a particular reason for pushing this new system?
I think we should focus on keeping things very simple.

File: cmd/genji/shell/command.go [hunk 9]
```diff
@@ -92,16 +101,97 @@ func runIndexesCmd(db *genji.DB, in []string) error {
...
+	cmdList := make([]string, len(commands))
+	for i := range commands {
+		cmdList[i] = commands[i].Name
+	}
+
+	return cmdList
+}
+
+// getCommandSuggestion returns the suggested command.
+func getCommandSuggestion(in string) Command {
```
Comment (asdine) [inline cmd/genji/shell/command.go:147]: This doesn't do what the comment suggests, it creates a command that itself determines which command to run. I think this is a bit hacky.

File: cmd/genji/shell/command.go (hunk 5 above)
Comment (asdine) [reply cmd/genji/shell/command.go:119]: This is marked as resolved but the database is not closed, isn't it?

File: cmd/genji/shell/command.go [hunk 10]
```diff
@@ -2,20 +2,30 @@ package shell
...
-	".help":                 "\t\t\tList all commands.",
+type Command struct {
+	Name    string
+	Options string
+	Help    string
+	Action  func(db *genji.DB, cmd []string) error
 }
 
-func runTablesCmd(db *genji.DB, cmd []string) error {
+const TotalCommands = 4
```
Comment (asdine) [inline cmd/genji/shell/command.go:19]: Creating globals must be done for a very good reason in my opinion, especially when they are exported.

**Review by asdine — CHANGES_REQUESTED** [review]
I don't think this new `Command` type is worth the trouble, when a simple switch case like we currently have does the job. I believe this PR should simply focus on adding the command suggestions and not redesigning the system, unless there is a very good reason.

File: cmd/genji/shell/command.go [hunk 11]
```diff
@@ -2,20 +2,13 @@ package shell
...
 
-var commands = map[string]string{
-	".tables":               "\t\tList names of tables.",
-	".exit":                 "\t\t\tExit this program.",
-	".indexes [table_name]": "\tDisplay all indexes or the indexes of the given table name.",
-	".help":                 "\t\t\tList all commands.",
-}
-
-func runTablesCmd(db *genji.DB, cmd []string) error {
+// displayTables show all tables.
```
Comment (asdine) [inline cmd/genji/shell/command.go:10]:
```suggestion
// displayTables shows all tables.
```

File: cmd/genji/shell/command.go [hunk 12]
```diff
@@ -92,17 +85,58 @@ func runIndexesCmd(db *genji.DB, in []string) error {
 	return fmt.Errorf("usage: .indexes [tablename]")
 }
 
-// runHelpCmd display all available dot commands.
-func runHelpCmd() error {
-	var keys []string
-	for k := range commands {
-		keys = append(keys, k)
+// displayHelp show all available cmd.
```
Comment (asdine) [inline cmd/genji/shell/command.go:88]:
```suggestion
// displayHelp shows all available commands.
```

File: cmd/genji/shell/command.go [hunk 13]
```diff
@@ -92,17 +85,58 @@ func runIndexesCmd(db *genji.DB, in []string) error {
...
+func displayHelp() error {
+
```
Comment (asdine) [inline cmd/genji/shell/command.go:90]:
```suggestion
```

File: cmd/genji/shell/command.go [hunk 14]
```diff
@@ -92,17 +85,58 @@ func runIndexesCmd(db *genji.DB, in []string) error {
...
+		indent := spaces - len(commands[i].options) - len(commands[i].name)
+		fmt.Printf("%s %s%*s%s\n", commands[i].name, commands[i].options, indent, "", commands[i].help)
+	}
+
+	return nil
+}
+
+// displaySuggestion shows suggestions.
+func displaySuggestion(in string) error {
+
```
Comment (asdine) [inline cmd/genji/shell/command.go:117]:
```suggestion
```

File: cmd/genji/shell/command.go [hunk 15]
```diff
@@ -92,17 +85,58 @@ func runIndexesCmd(db *genji.DB, in []string) error {
...
+	var suggestions []string
+	commands := []string{".tables", ".indexes", ".help", ".exit"}
+	for _, c := range commands {
+		d := levenshtein.ComputeDistance(c, in)
+		// input should be equal to half command name to get a suggestion.
```
Comment (asdine) [inline cmd/genji/shell/command.go:122]:
```suggestion
		// input should be equal to half the command name to get a suggestion.
```

File: cmd/genji/shell/command.go [hunk 16]
```diff
@@ -92,17 +85,58 @@ func runIndexesCmd(db *genji.DB, in []string) error {
...
 
-	sort.Strings(keys)
-	for _, k := range keys {
-		fmt.Printf("%s %s\n", k, commands[k])
+	fmt.Printf("\"%s\" is not a command. Did you mean: ", in)
+	for i := range suggestions {
+		if i > 0 {
+			fmt.Printf(", ")
+		}
+		fmt.Printf("\"%s\"", suggestions[i])
```
Comment (asdine) [inline cmd/genji/shell/command.go:137]:
```suggestion
		fmt.Printf("%q", suggestions[i])
```

File: cmd/genji/shell/command.go [hunk 17]
```diff
@@ -2,20 +2,13 @@ package shell
...
 import (
 	"fmt"
-	"sort"
-
+	"github.com/agnivade/levenshtein"
 	"github.com/genjidb/genji"
 	"github.com/genjidb/genji/document"
 )
 
-var commands = map[string]string{
```
Comment (asdine) [inline cmd/genji/shell/command.go:11]: Before, we had one place where we kept the commands, now we have 3 places we need to update to add them. Why not keeping this map and extract + sort the keys whenever we need them?

File: cmd/genji/shell/command.go [hunk 18]
```diff
@@ -92,17 +85,58 @@ func runIndexesCmd(db *genji.DB, in []string) error {
...
+	commands = append(commands, command{name: ".indexes",
+		options: "[table name]",
+		help:    "Display all indexes or the indexes of the given table name."})
+
+	commands = append(commands, command{name: ".tables", help: "List names of tables."})
+	commands = append(commands, command{name: ".help", help: "Show available cmd."})
+	commands = append(commands, command{name: ".exit", help: "Exit this program."})
+
+	for i := range commands {
+		// spaces indentation to display helps aligned.
```
Comment (asdine) [inline cmd/genji/shell/command.go:106]:
```suggestion
		// spaces indentation for readability.
```

File: cmd/genji/shell/shell.go [hunk 19]
```diff
@@ -370,6 +384,16 @@ func (sh *Shell) getAllTables() ([]string, error) {
 }
 
 func (sh *Shell) completer(in prompt.Document) []prompt.Suggest {
+
```
Comment (asdine) [inline cmd/genji/shell/shell.go:387]:
```suggestion
```

Starting a function with a blank line is evil

File: cmd/genji/shell/command.go [hunk 20]
```diff
@@ -92,17 +85,58 @@ func runIndexesCmd(db *genji.DB, in []string) error {
...
+	for _, c := range commands {
+		d := levenshtein.ComputeDistance(c, in)
+		// input should be equal to half command name to get a suggestion.
+		if d < (len(c) / 2) {
+			suggestions = append(suggestions, c)
+		}
+	}
+
+	if len(suggestions) == 0 {
+		return fmt.Errorf("unknown command %q. Enter \".help\" for help.", in)
```
Comment (asdine) [inline cmd/genji/shell/command.go:129]:
```suggestion
		return fmt.Errorf("Unknown command %q. Enter \".help\" for help.", in)
```

**Review by asdine — CHANGES_REQUESTED** [review]
I think it could be even simpler, but we're almost there 👍

File: cmd/genji/shell/command.go [hunk 21]
```diff
@@ -2,20 +2,21 @@ package shell
...
 )
 
 var commands = map[string]string{
-	".tables":               "\t\tList names of tables.",
-	".exit":                 "\t\t\tExit this program.",
-	".indexes [table_name]": "\tDisplay all indexes or the indexes of the given table name.",
-	".help":                 "\t\t\tList all commands.",
+	".exit":                 "exit this program",
+	".tables":               "shows all tables",
+	".indexes [table_name]": "shows all indexes",
```
Comment (asdine) [inline cmd/genji/shell/command.go:14]: Is there a reason why we are changing the command descriptions?

File: cmd/genji/shell/command.go [hunk 22]
```diff
@@ -2,20 +2,21 @@ package shell
...
+	".help":                 "shows help",
 }
 
-func runTablesCmd(db *genji.DB, cmd []string) error {
+// displayTables shows all tables.
+func displayTables(db *genji.DB, cmd []string) error {
```
Comment (asdine) [inline cmd/genji/shell/command.go:19]: I don't think there is a need to rename this function: we are running the tables command, this is not only displaying something.
Think of it this way: everything the user runs that starts with `.` is a command. However, displaying the suggestion is not a command, that's why we named that function `displaySuggestion`.

File: cmd/genji/shell/command.go [hunk 23]
```diff
@@ -92,17 +93,49 @@ func runIndexesCmd(db *genji.DB, in []string) error {
 	return fmt.Errorf("usage: .indexes [tablename]")
 }
 
-// runHelpCmd display all available dot commands.
-func runHelpCmd() error {
-	var keys []string
+// displayHelp shows all available commands.
+func displayHelp() error {
```
Comment (asdine) [inline cmd/genji/shell/command.go:97]: Same as above

File: cmd/genji/shell/command.go [hunk 24]
```diff
@@ -92,17 +93,49 @@ func runIndexesCmd(db *genji.DB, in []string) error {
...
+
+	return nil
+}
+
+// displaySuggestion shows suggestions.
+func displaySuggestion(in string) error {
+	var suggestions []string
+	for _, c := range commands {
+		d := levenshtein.ComputeDistance(c, in)
+		// input should be equal to the half command name to get a suggestion.
```
Comment (asdine) [inline cmd/genji/shell/command.go:120]:
```suggestion
		// input should be at least half the command name to get a suggestion.
```

File: cmd/genji/shell/shell.go [hunk 25]
```diff
@@ -123,6 +125,16 @@ func Run(opts *Options) error {
 	return sh.dumpHistory()
 }
 
+func (sh *Shell) registerCommandSuggestion() {
```
Comment (asdine) [inline cmd/genji/shell/shell.go:128]: I don't think we need to register this and keep the state. We should simply rename it `getCommandSuggestions` and return the suggestions based on the `commands` map.
The `expected` slice below is duplicating the commands information which is not ideal, let's just use the global one.

**Review by asdine — CHANGES_REQUESTED** [review]
This is better! Just a few comments, especially regarding renaming decisions. It's definitely simpler like this!

File: cmd/genji/shell/shell.go [hunk 26]
```diff
@@ -370,6 +385,20 @@ func (sh *Shell) getAllTables() ([]string, error) {
 }
 
 func (sh *Shell) completer(in prompt.Document) []prompt.Suggest {
+	if strings.HasPrefix(in.Text, ".") {
+		// This check avoids infinite suggestions.
+		if isCommand(in.Text) {
```
Comment (asdine) [inline cmd/genji/shell/shell.go:390]: This is necessary because you are using `GetWordBeforeCursor` below. If you change it to `in.Text` instead you won't have this problem.

File: document/array.go [hunk 27]
```diff
@@ -8,7 +8,7 @@ import (
 // ErrValueNotFound must be returned by Array implementations, when calling the GetByIndex method and
 // the index wasn't found in the array.
 var (
-	ErrValueNotFound   = errors.New("value not found")
```
Comment (asdine) [inline document/array.go:11]: Please avoid changes that are not related to the PR, this can make rebases more difficult for everyone

File: document/document_test.go [hunk 28]
```diff
@@ -92,9 +92,9 @@ func TestFieldBuffer(t *testing.T) {
 			{"add field", `{"a": {"b": [1, 2, 3]}}`, `c`, document.NewTextValue("foo"), `{"a": {"b": [1, 2, 3]}, "c": "foo"}`, false},
 			{"nested doc", `{"a": "foo"}`, `a`, document.NewDocumentValue(document.NewFieldBuffer().
 				Add("b", document.NewArrayValue(document.NewValueBuffer().
-				Append(document.NewIntegerValue(1)).
-				Append(document.NewIntegerValue(2)).
-				Append(document.NewIntegerValue(3))))), `{"a": {"b": [1, 2, 3]}}`, false},
+					Append(document.NewIntegerValue(1)).
```
Comment (asdine) [inline document/document_test.go:95]: Same here

**Review by asdine — CHANGES_REQUESTED** [review]
A few comments, but we're almost there 👍🏼

**Comment by asdine** [comment]
Also, please rebase your PR with master, otherwise I can't merge it

File: cmd/genji/shell/shell.go [hunk 29]
```diff
@@ -122,7 +126,24 @@ func Run(opts *Options) error {
 	)
 
 	e.Run()
-	return nil
+
```
Comment (asdine) [inline cmd/genji/shell/shell.go:129]:
```suggestion
```

File: cmd/genji/shell/shell.go [hunk 30]
```diff
@@ -122,7 +126,24 @@ func Run(opts *Options) error {
...
+		err = sh.db.Close()
+		if err != nil {
+			return err
+		}
+	}
+
+	return sh.dumpHistory()
+}
+
+func (sh *Shell) loadCommandSuggestion()  {
```
Comment (asdine) [inline cmd/genji/shell/shell.go:141]:
```suggestion
func (sh *Shell) loadCommandSuggestions()  {
```

File: cmd/genji/shell/command.go [hunk 31]
```diff
@@ -92,17 +97,42 @@ func runIndexesCmd(db *genji.DB, in []string) error {
...
 	}
 
-	sort.Strings(keys)
-	for _, k := range keys {
-		fmt.Printf("%s %s\n", k, commands[k])
+	return nil
+}
+
+// displaySuggestion shows suggestions.
+func displaySuggestion(in string) error {
```
Comment (asdine) [inline cmd/genji/shell/command.go:113]:
```suggestion
func displaySuggestions(in string) error {
```

**Review by asdine — APPROVED** [review]
Some final typos and spacing issues, otherwise everything is good now

**Review by asdine — APPROVED** [review]
Thank you @tzzed !
